File: clients/java/src/main/java/com/zeeplabs/orbit/FilesClient.java
package com.zeeplabs.orbit;

import com.zeeplabs.orbit.Types.FileResponse;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.UUID;

public class FilesClient {

    private final OrbitClient client;
    private final ObjectMapper mapper = new ObjectMapper();

    FilesClient(OrbitClient client) {
        this.client = client;
    }

    private String path(String p) {
        return "files/" + p.replaceAll("^/", "");
    }

    public FileResponse upload(String filename, byte[] data, String mimeType) throws Exception {
        String boundary = UUID.randomUUID().toString();
        String header = "--" + boundary + "\r\n"
                + "Content-Disposition: form-data; name=\"file\"; filename=\"" + filename + "\"\r\n"
                + "Content-Type: " + mimeType + "\r\n\r\n";
        String footer = "\r\n--" + boundary + "--\r\n";

        byte[] headerBytes = header.getBytes(StandardCharsets.UTF_8);
        byte[] footerBytes = footer.getBytes(StandardCharsets.UTF_8);
        byte[] body = new byte[headerBytes.length + data.length + footerBytes.length];
        System.arraycopy(headerBytes, 0, body, 0, headerBytes.length);
        System.arraycopy(data, 0, body, headerBytes.length, data.length);
        System.arraycopy(footerBytes, 0, body, headerBytes.length + data.length, footerBytes.length);

        URL url = new URI(client.url(path(""))).toURL();
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Authorization", "Bearer " + client.config.jwt);
        conn.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);
        conn.setDoOutput(true);

        try (OutputStream os = conn.getOutputStream()) {
            os.write(body);
        }

        int status = conn.getResponseCode();
        byte[] resp = (status >= 400 ? conn.getErrorStream() : conn.getInputStream()).readAllBytes();
        String raw = new String(resp, StandardCharsets.UTF_8);

        if (status >= 400) {
            Map<String, Object> err = mapper.readValue(raw, Map.class);
            throw new RuntimeException((String) err.getOrDefault("error", "HTTP " + status));
        }

        return mapper.readValue(raw, FileResponse.class);
    }

    public FileResponse[] list(int limit, int offset) {
        return client.request("GET", path("?limit=" + limit + "&offset=" + offset), null, FileResponse[].class);
    }

    public FileResponse get(String id) {
        return client.request("GET", path(id), null, FileResponse.class);
    }

    public void delete(String id) {
        client.request("DELETE", path(id), null, null);
    }

    @SuppressWarnings("unchecked")
    public String signedURL(String id, int ttl) {
        Map<String, Object> resp = client.request("GET", path(id + "/url?ttl=" + ttl), null, Map.class);
        return (String) resp.get("url");
    }
}

File: clients/php/src/FilesClient.php
<?php

namespace Zeeplabs\Orbit;

class FilesClient
{
    public function __construct(
        private readonly OrbitClient $client
    ) {}

    private function path(string $p): string
    {
        return 'files/' . ltrim($p, '/');
    }

    public function upload(string $filename, string $data, string $mimeType): array
    {
        $boundary = bin2hex(random_bytes(16));
        $header = "--{$boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"{$filename}\"\r\nContent-Type: {$mimeType}\r\n\r\n";
        $footer = "\r\n--{$boundary}--\r\n";
        $body = $header . $data . $footer;

        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => $this->client->url($this->path('')),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => $body,
            CURLOPT_HTTPHEADER => [
                'Authorization: Bearer ' . $this->client->jwt,
                'Content-Type: multipart/form-data; boundary=' . $boundary,
            ],
        ]);

        $response = curl_exec($ch);
        $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        $data = json_decode($response, true);

        if ($status >= 400) {
            $msg = $data['error'] ?? 'HTTP ' . $status;
            throw new OrbitException($msg, $status);
        }

        return $data;
    }

    public function list(int $limit = 50, int $offset = 0): array
    {
        return $this->client->request('GET', $this->path("?limit={$limit}&offset={$offset}"));
    }

    public function get(string $id): array
    {
        return $this->client->request('GET', $this->path($id));
    }

    public function delete(string $id): void
    {
        $this->client->request('DELETE', $this->path($id));
    }

    public function signedURL(string $id, int $ttl = 3600): string
    {
        $resp = $this->client->request('GET', $this->path("{$id}/url?ttl={$ttl}"));
        return $resp['url'];
    }
}
